stubs/WebOb/webob/dec.pyi

```diff
@@ -1,49 +1,196 @@
-from _typeshed import Incomplete
-from _typeshed.wsgi import WSGIApplication
-from typing import ClassVar, overload
-from typing_extensions import Self
+from _typeshed.wsgi import StartResponse, WSGIApplication, WSGIEnvironment
+from collections.abc import Callable, Iterable, Mapping
+from typing import Any, Generic, TypeVar, overload
+from typing_extensions import Concatenate, Never, ParamSpec, Self, TypeAlias
 
 from webob.request import Request
+from webob.response import Response
 
-class wsgify:
-    RequestClass: ClassVar[type[Request]]
-    func: Incomplete
-    args: Incomplete
-    kwargs: Incomplete
-    middleware_wraps: Incomplete
+_AnyResponse: TypeAlias = Response | WSGIApplication | str | None
+_S = TypeVar("_S")
+_AppT = TypeVar("_AppT", bound=WSGIApplication)
+_AppT_contra = TypeVar("_AppT_contra", bound=WSGIApplication, contravariant=True)
+_RequestT = TypeVar("_RequestT", bound=Request)
+_RequestT_contra = TypeVar("_RequestT_contra", bound=Request, contravariant=True)
+_P = ParamSpec("_P")
+_P2 = ParamSpec("_P2")
+
+_RequestHandlerCallable: TypeAlias = Callable[Concatenate[_RequestT_contra, _P], _AnyResponse]
+_RequestHandlerMethod: TypeAlias = Callable[Concatenate[Any, _RequestT_contra, _P], _AnyResponse]
+_MiddlewareCallable: TypeAlias = Callable[Concatenate[_RequestT_contra, _AppT_contra, _P], _AnyResponse]
+_MiddlewareMethod: TypeAlias = Callable[Concatenate[Any, _RequestT_contra, _AppT_contra, _P], _AnyResponse]
+_RequestHandler: TypeAlias = _RequestHandlerCallable[_RequestT_contra, _P] | _RequestHandlerMethod[_RequestT_contra, _P]
+_Middleware: TypeAlias = (
+    _MiddlewareCallable[_RequestT_contra, _AppT_contra, _P] | _MiddlewareMethod[_RequestT_contra, _AppT_contra, _P]
+)
+
+class wsgify(Generic[_RequestT_contra, _P]):
+    RequestClass: type[Request]
+    func: _RequestHandler[_RequestT_contra, _P] | None
+    args: tuple[Any, ...]
+    kwargs: dict[str, Any]
+    middleware_wraps: WSGIApplication | None
+    # NOTE: We disallow passing args/kwargs using this direct API, because
```

Comment:
Why can't we have an overload where args/kwargs are nonempty and the rest of the types reduce to Any? That way it's only nonsafe if the user explicitly passed args/kwargs.

Reply:
I tried that (with an absolutely absurd number of overloads), unfortunately, when you do that it will complain if you use it to decorate a callable that doesn't also itself take arbitrary `args` and `kwargs`. The types of the arguments don't matter in this case, but the number of arguments unfortunately does matter.

i.e. stuff like this won't work:
```
@wsgify(args=(1.0,))  # type error: wsgify[Request, ...].__call__ no overloads match
def foo(req: Request, timeout: float) -> None: ...
```

I could maybe get around this by adding two more subclasses, which are generic in `_F` instead of `_P`, so they can decorate arbitrary callables and adding a `__new__` to return them, but it doesn't really seem worth the effort, since it won't help with subclasses of wsgify.

The alternative is giving up an parameter safety altogether, or try to only use a `ParamSpec` for the `middleware` call.